mllib/src/main/scala/org/apache/spark/mllib/rdd/SlidingRDD.scala

```diff
@@ -66,36 +69,54 @@ class SlidingRDD[T: ClassTag](@transient val parent: RDD[T], val windowSize: Int
     if (n == 0) {
       Array.empty
     } else if (n == 1) {
-      Array(new SlidingRDDPartition[T](0, parentPartitions(0), Seq.empty))
+      Array(new SlidingRDDPartition[T](0, parentPartitions(0), Seq.empty, 0))
     } else {
       val n1 = n - 1
-      val w1 = windowSize - 1
-      // Get the first w1 items of each partition, starting from the second partition.
-      val nextHeads =
-        parent.context.runJob(parent, (iter: Iterator[T]) => iter.take(w1).toArray, 1 until n)
+      // Get partitions sizes
+      val sizes =
+        parent.context.runJob(parent, (iter: Iterator[T]) => iter.length, 0 until n)
```

Comment:
If `step > 1`, in a single Spark job, we can collect the following:

1. the first `w1` elements from each partition (except the first one)
2. the size of each partition

Then we can compute the offset for each partition.